### PR TITLE
CB-17216 Integrate consumption into environment (create & terminate)

### PR DESCRIPTION
**Pull request by lajosrodek**
* Add `cloud-consumption-api` as new compile dependency for `environment` service.
* Integrate consumption schedule / unschedule invocations into environment create & terminate flows.
* Add new environment states and related notification messages.
* Some code cleanup.
* Testing:
  * Manual testing in local CB.
  * Added new UT and updated existing ones.

### CODE REVIEW

**Review by topolyai5 — APPROVED** [review]

File: environment/src/main/java/com/sequenceiq/environment/environment/flow/creation/handler/StorageConsumptionCollectionSchedulingHandler.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    private void scheduleStorageConsumptionCollection(EnvironmentDto environmentDto) {
+        StorageConsumptionRequest request = new StorageConsumptionRequest();
+        String resourceCrn = environmentDto.getResourceCrn();
+        request.setEnvironmentCrn(resourceCrn);
+        request.setMonitoredResourceCrn(resourceCrn);
+        request.setMonitoredResourceName(environmentDto.getName());
+        request.setMonitoredResourceType(ResourceType.ENVIRONMENT);
+        request.setStorageLocation(environmentDto.getTelemetry().getLogging().getStorageLocation());
```
Comment (bergerdenes) [inline environment/src/main/java/com/sequenceiq/environment/environment/flow/creation/handler/StorageConsumptionCollectionSchedulingHandler.java:91]: What if telemetry is null? (Of course this is a test edge case only but for example I use to create my environments without telemetry logging)

A second question: what about backup storage location?

File: environment/src/main/java/com/sequenceiq/environment/environment/flow/deletion/handler/StorageConsumptionCollectionUnschedulingHandler.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import reactor.bus.Event;
+
+@Component
+public class StorageConsumptionCollectionUnschedulingHandler extends EventSenderAwareHandler<EnvironmentDeletionDto> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(StorageConsumptionCollectionUnschedulingHandler.class);
+
+    private final EnvironmentService environmentService;
+
+    private final HandlerExceptionProcessor exceptionProcessor;
```
Comment (bergerdenes) [inline environment/src/main/java/com/sequenceiq/environment/environment/flow/deletion/handler/StorageConsumptionCollectionUnschedulingHandler.java:29]: This component is not used in the counterpart `SchedulingHandler`

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/deletion/EnvDeleteActionsTest.java [hunk 3]
```diff
@@ -178,52 +178,36 @@ void setUp() {
     @Test
     void idbmmsDeleteActionTestHappyPath() {
         testDeleteActionHappyPath(underTest::idbmmsDeleteAction, DELETE_IDBROKER_MAPPINGS_EVENT.selector(),
-                EnvironmentStatus.IDBROKER_MAPPINGS_DELETE_IN_PROGRESS, ResourceEvent.ENVIRONMENT_IDBROKER_MAPPINGS_DELETION_STARTED);
-    }
-
-    @Test
```
Comment (bergerdenes) [inline environment/src/test/java/com/sequenceiq/environment/environment/flow/deletion/EnvDeleteActionsTest.java:184]: Why is the deletion?

**Review by bergerdenes — CHANGES_REQUESTED** [review]

File: common/src/main/java/com/sequenceiq/cloudbreak/event/ResourceEvent.java [hunk 4]
```diff
@@ -98,6 +100,7 @@ public enum ResourceEvent {
     ENVIRONMENT_CLUSTER_DEFINITION_DELETE_STARTED("environment.clusterdefinition.deletion.started"),
     ENVIRONMENT_DATABASE_DELETION_STARTED("environment.database.deletion.started"),
     ENVIRONMENT_FREEIPA_DELETION_STARTED("environment.freeipa.deletion.started"),
+    ENVIRONMENT_STORAGE_CONSUMPTION_COLLECTION_UNSCHEDULING_STARTED("environment.storage.consumption.collection.unscheduling.started"),
```
[the remark this answers is not in the document]
Comment (lajosrodek) [reply common/src/main/java/com/sequenceiq/cloudbreak/event/ResourceEvent.java:103]: Sorry, I do not get it. This is a notification about the start event.

File: environment/src/main/java/com/sequenceiq/environment/environment/flow/creation/handler/StorageConsumptionCollectionSchedulingHandler.java (hunk 1 above)
Comment (lajosrodek) [reply environment/src/main/java/com/sequenceiq/environment/environment/flow/creation/handler/StorageConsumptionCollectionSchedulingHandler.java:91]: Good points!
* The log location is mandatory for env creation. It is a required parameter for CDP CLI, and also a mandatory input field in the UI wizard. I wonder if this can ever be `null`. Perhaps only for `MOCK` or `YARN` providers? I have no idea TBH.
* The backup storage location is an interesting one. I am not sure we touched upon that if it also needs to be measured.

@nadamdb Could you please chime in regarding the second bullet point? :smile:

Comment (lajosrodek) [reply environment/src/main/java/com/sequenceiq/environment/environment/flow/creation/handler/StorageConsumptionCollectionSchedulingHandler.java:91]: One note: The backup location defaults to the log location unless overridden. So if we were to decide on reporting consumption for the backup location as well, the schedule & unschedule calls would have to take this into consideration (and skip the backup location if it matches the log one).

File: environment/src/main/java/com/sequenceiq/environment/environment/flow/deletion/handler/StorageConsumptionCollectionUnschedulingHandler.java (hunk 2 above)
Comment (lajosrodek) [reply environment/src/main/java/com/sequenceiq/environment/environment/flow/deletion/handler/StorageConsumptionCollectionUnschedulingHandler.java:29]: It is only needed for the error handling in the delete flow handlers. The layout / approach is a direct copy paste of other handlers around.

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/deletion/EnvDeleteActionsTest.java (hunk 3 above)
Comment (lajosrodek) [reply environment/src/test/java/com/sequenceiq/environment/environment/flow/deletion/EnvDeleteActionsTest.java:184]: Those test cases were invalid, they were supposed to verify a scenario that does not apply (any more) to these delete actions. In particular, there is no distinction regarding the "environment missing" case in the action logic. (This is in direct opposition with the create actions that do differentiate between environment absent & present.)

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/deletion/EnvDeleteActionsTest.java [hunk 5]
```diff
@@ -178,52 +178,36 @@ void setUp() {
...
+                EnvDeleteState.IDBROKER_MAPPINGS_DELETE_STARTED_STATE);
     }
 
     @Test
     void publicKeyDeleteActionTestHappyPath() {
         testDeleteActionHappyPath(underTest::publickeyDeleteAction, DELETE_PUBLICKEY_EVENT.selector(),
-                EnvironmentStatus.PUBLICKEY_DELETE_IN_PROGRESS, ResourceEvent.ENVIRONMENT_PUBLICKEY_DELETION_STARTED);
-    }
-
-    @Test
```
[the remark this answers is not in the document]
Comment (lajosrodek) [reply environment/src/test/java/com/sequenceiq/environment/environment/flow/deletion/EnvDeleteActionsTest.java:195]: Like above. :smile:

File: common/src/main/java/com/sequenceiq/cloudbreak/event/ResourceEvent.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (bergerdenes) [reply common/src/main/java/com/sequenceiq/cloudbreak/event/ResourceEvent.java:103]: ok

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/deletion/EnvDeleteActionsTest.java (hunk 3 above)
Comment (bergerdenes) [reply environment/src/test/java/com/sequenceiq/environment/environment/flow/deletion/EnvDeleteActionsTest.java:184]: ok

File: environment/src/test/java/com/sequenceiq/environment/environment/flow/deletion/EnvDeleteActionsTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (bergerdenes) [reply environment/src/test/java/com/sequenceiq/environment/environment/flow/deletion/EnvDeleteActionsTest.java:195]: ok

File: environment/src/main/java/com/sequenceiq/environment/environment/flow/creation/handler/StorageConsumptionCollectionSchedulingHandler.java (hunk 1 above)
Comment (nadamdb) [reply environment/src/main/java/com/sequenceiq/environment/environment/flow/creation/handler/StorageConsumptionCollectionSchedulingHandler.java:91]: Just to make a note about it here: as discussed, backup location scheduling will be added later as that location is not static and can be changed after env creation so we need a bit different approach there. 
The scheduling would work similarly to log location scheduling but the unscheduling is a bit trickier and we have to prevent accidentally unscheduling the log location in cases where the backup bucket was originally the same as the log one but is later changed to a different one (as the crn of the monitored resource and the storage location together uniquely identify a consumption collection, hence if both are the same we only schedule it once).

File: environment/src/main/resources/application.yml [hunk 6]
```diff
@@ -93,6 +93,10 @@ environment:
   sdx:
     url: http://localhost:8086
     contextPath: /dl
+  consumption:
+    url: http://localhost:8099
+    contextPath: /consumption
+    enabled: false
```
Comment (nadamdb) [inline environment/src/main/resources/application.yml:99]: Is it intentionally disabled?

**Review by nadamdb — COMMENTED** [review]
Just had a few questions regarding configs, otherwise LGTM

File: environment/build.gradle [hunk 7]
```diff
@@ -67,10 +67,10 @@ dependencies {
     implementation     project(":cloud-mock")
     implementation     project(":cloud-yarn")
     implementation     project(":cloud-gcp")
-    implementation     project(":core-api")
```
[the remark this answers is not in the document]
Comment (lajosrodek) [reply environment/build.gradle:70]: Yes. The same dependency already appears in line 63.

File: environment/src/main/java/com/sequenceiq/environment/configuration/api/ServiceEndpointConfig.java [hunk 8]
```diff
@@ -57,9 +57,18 @@ public class ServiceEndpointConfig {
...
 
-    @Value("${environment.sdx.server.contextPath:/dl}")
+    @Value("${environment.sdx.contextPath}")
     private String sdxRootContextPath;
 
+    @Value("${environment.consumption.url:}")
+    private String consumptionServiceUrl;
+
+    @Value("${environment.consumption.serviceId:}")
+    private String consumptionServiceId;
```
[the remark this answers is not in the document]
Comment (lajosrodek) [reply environment/src/main/java/com/sequenceiq/environment/configuration/api/ServiceEndpointConfig.java:67]: Like above, I followed the pattern used in this class for all the other services.

File: environment/src/main/java/com/sequenceiq/environment/environment/flow/creation/handler/StorageConsumptionCollectionSchedulingHandler.java (hunk 1 above)
Comment (lajosrodek) [reply environment/src/main/java/com/sequenceiq/environment/environment/flow/creation/handler/StorageConsumptionCollectionSchedulingHandler.java:91]: I will add support for the backup location scheduling/unscheduling now, but full support for the backup location edit scenario will be tackled later in [CB-17410](https://jira.cloudera.com/browse/CB-17410).

File: environment/src/main/resources/application.yml (hunk 6 above)
Comment (nadamdb) [reply environment/src/main/resources/application.yml:99]: Thanks for clarifying, it was my guess too that it's mainly for local setups, just wanted to be sure.

**Review by bergerdenes — APPROVED** [review]

**Review by nadamdb — APPROVED** [review]
LGTM

**Review by biharitomi — APPROVED** [review]
